File: trunk/QTwikeAnalyzer/src/energex/gui/SettingsDialog.java
package energex.gui;

import java.util.Enumeration;
import com.trolltech.qt.core.QSettings;
import com.trolltech.qt.gui.QComboBox;
import com.trolltech.qt.gui.QDialog;

import energex.gui.Ui_Settings;
import gnu.io.CommPortIdentifier;

public class SettingsDialog extends QDialog {
	
	Ui_Settings ui = new Ui_Settings();
	QSettings settings = null;
	
	public SettingsDialog() {
		ui.setupUi(this);	
	}
	
	@SuppressWarnings("unchecked")
	public SettingsDialog(TwikeAnalyzer parent) {
		super(parent);
		ui.setupUi(this);
		
		ui.comboSerial.setInsertPolicy(QComboBox.InsertPolicy.InsertAlphabetically);
		
		// Get an enumeration of all ports known to JavaComm
		//
		Enumeration portIdentifiers = CommPortIdentifier.getPortIdentifiers();
		
		// Check each port identifier if 
		//   (a) it indicates a serial (not a parallel) port, and
		//   (b) matches the desired name.
		//
		while (portIdentifiers.hasMoreElements())
		{
		    CommPortIdentifier pid = (CommPortIdentifier) portIdentifiers.nextElement();
		    if(pid.getPortType() == CommPortIdentifier.PORT_SERIAL ) {
		    	ui.comboSerial.addItem(pid.getName(), pid);
		    }
		}
		
		// restore current default
		settings = new QSettings();
		String storedPort = (String) settings.value("Port");

		if(storedPort != null) {
			int idx = ui.comboSerial.findText(storedPort);
			ui.comboSerial.setCurrentIndex(idx);
		}
	}
	
	public void on_buttonBox_accepted() {
		// store selection
		settings.setValue("Port", ui.comboSerial.currentText());
	}
	
	public String getPortName() {
		return ui.comboSerial.currentText();
	}
	
	public CommPortIdentifier getPort() {
    	int idx = ui.comboSerial.currentIndex();
    	return (CommPortIdentifier) ui.comboSerial.itemData(idx);
    }
}

File: QTwikeAnalyzer/settingsdialog.cpp
#include "settingsdialog.h"
#include <QSettings>

#include "ui_settingsdialog.h"

SettingsDialog::SettingsDialog(QWidget *parent) :
    QDialog(parent),
    m_ui(new Ui::SettingsDialog)
{
    m_ui->setupUi(this);

    m_ui->comboSerial->setInsertPolicy(QComboBox::InsertAlphabetically);

    m_ui->comboSerial->addItem("/dev/ttyUSB0");

    // restore current default
    QSettings settings;
    settings.beginGroup("Settings");
    QString storedPort = settings.value("Port").toString();
    settings.endGroup();

    if(!storedPort.isEmpty())
    {
        int idx = m_ui->comboSerial->findText(storedPort);
        m_ui->comboSerial->setCurrentIndex(idx);
    }
}

SettingsDialog::~SettingsDialog()
{
    delete m_ui;
}

void SettingsDialog::on_buttonBox_accepted()
{
    // store selection
    QSettings settings;

    settings.beginGroup("Settings");
    settings.setValue("Port", m_ui->comboSerial->currentText());
    settings.endGroup();
}

QString SettingsDialog::getPortName()
{
    return m_ui->comboSerial->currentText();
}

void SettingsDialog::changeEvent(QEvent *e)
{
    QDialog::changeEvent(e);
    switch (e->type()) {
    case QEvent::LanguageChange:
        m_ui->retranslateUi(this);
        break;
    default:
        break;
    }
}
